oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2DeviceVerificationAuthenticationProvider.java

```diff
@@ -109,6 +109,21 @@ public Authentication authenticate(Authentication authentication) throws Authent
 			this.logger.trace("Retrieved authorization with user code");
 		}
 
+		OAuth2Authorization.Token<OAuth2UserCode> userCode = authorization.getToken(OAuth2UserCode.class);
+		if (userCode.isInvalidated()) {
+			if (this.logger.isTraceEnabled()) {
+				this.logger.trace("User code is invalided");
+			}
+			throw new OAuth2AuthenticationException(OAuth2ErrorCodes.INVALID_GRANT);
+		}
+
+		if (userCode.isExpired()) {
+			if (this.logger.isTraceEnabled()) {
+				this.logger.trace("User code is expired");
+			}
+			throw new OAuth2AuthenticationException(OAuth2ErrorCodes.INVALID_GRANT);
```

Comment:
Is it worth providing more context on the error..
```suggestion
						OAuth2Error error = new OAuth2Error(OAuth2ErrorCodes.INVALID_GRANT, "User code is expired.", null);
						throw new OAuth2AuthenticationException(error);
```